### ISSUE TITLE
Local scope of scripts in bin directory

### ISSUE THREAD

**Issue by phiweger** [nextflow-io/nextflow#1798]
## New feature

When we import a subworkflow into a larger one, and this subwf uses scripts (like some Python parser), then currently this script needs to be copied from the subwf's bin directory to the larger wf's bin directory. However, wouldn't it me more logical for the subwf to look if it can find the script "locally" first?

**Comment by abhi18av** [comment]
I guess that, in the same vein, same update could be made for the script `templates` as well.

**Comment by grst** [comment]
This feature would be super helpful for [nf-core/modules](https://github.com/nf-core/modules)! 

To some degree, this is already possible with `$moduleDir`. It is possible to run a script from the module directory using 

```nextflow
script:
"""
${moduleDir}/bin/my_script.sh
"""
```

However, this breaks the caching functionality, i.e. modifying `my_script.sh` does not invalidate the cache. 

Moving forward, I see two options:
 * support a local `bin` directory that behaves the same way as the global one
 * consider files from `$moduleDir` when calculating the hash. 

I'm slightly in favor of the latter option, because
 * it is explicit, and avoids confusion and namespace clashes with the global bin dir
 * it does not only allow to get binaries, but also other resource files (e.g. report templates) from the module directory while being cache aware. 

---
CC @drpatelh, @ewels

**Comment by grst** [comment]
Another issue with `$moduleDir` is that it doesn't get mounted into docker containers. The scripts will, therefore, not be found when using the docker engine.

**Comment by drpatelh** [comment]
Yup, I agree it would be nice to eventually find a solution to this because it means custom modules that need to be shipped with their own executables can be packaged up and be self-contained in their own right too. @grst has created a proof of principle for us to be able to add such a [custom module](https://github.com/nf-core/modules/issues/294) to nf-core/modules but ran into the Docker mount path issue above. 

As @phiweger mentioned the only viable solution at present is to copy across the executable in the `bin/` directory of the main workflow for it be found correctly on `$PATH`.

**Comment by grst** [comment]
Would be great if at least the mounting-issue could be solved. The caching is less important as at least for nf-core pipelines the end user should never change the script anyway. 

I created a separate issue to track the mounting problem, including a reproducible example: https://github.com/nextflow-io/nextflow/issues/2240

**Comment by grst** [comment]
Before the stale-bot gets at this issue, I would want to throw in that the next nf-core hackathon is approaching and this issue will inevitably come up there again ;)

**Comment by ewels** [comment]
Bump

**Comment by nathaniel-mahieu** [comment]
Bump from an observer

As a note, another use for this functionality is the development of more complex custom code bases without having to rebuild docker or push/pull to github on every change.

**Comment by nathaniel-mahieu** [comment]
Not sure how this maps in all use cases but this pattern seems to be a suitable stand in for simple cases:

```
    path ASSETS, stageAs: 'assets' from "$projectDir/assets"
```

**Comment by ewels** [comment]
@nathaniel-mahieu - does that also work with `$moduleDir`? I guess so.. 🤔 That would be nicer in this context as we want it at module level rather than workflow, so that the scripts move with the modules when shared between workflows.

**Comment by phiweger** [comment]
@nathaniel-mahieu where do you declare this?

**Comment by nathaniel-mahieu** [comment]
@ewels - I am unsure if this generalizes to imports / modules.

@phiweger - in a process scope.

I have not tried this within a module. Sorry if I have confused the issue.

**Comment by nathaniel-mahieu** [comment]
The above example was a little terse so I made this pseudocode example that demonstrates it's usage. https://github.com/nathaniel-mahieu/nextflow-r-patterns

**Comment by AJnsm** [comment]
FWIW: the following seems to work for python scripts:

``` 
process myProcess {
    input:
    path myScript from "${projectDir}/myScript.py" 
    path utilities from "${projectDir}/utilities.py" 

    """
    python ${myScript}
    """

}
```

Where `myScript.py` contains 
```
from utilities import *
```

This brings `utilities.py` in scope. You'd have to to this separately for each process that needs access to the utilities module though.

**Comment by ewels** [comment]
Ah sorry I missed that. Is `$projectDir` not available within the module?

**Comment by pditommaso** [comment]
This has been solved via 77f55262 and c50e178f. 

The binary scripts must be placed in the module directory names `<module-dir>/resources/usr/bin`

    <module-dir>
        |-main.nf
        \-resources
            \-usr
                \-bin
                    |-your-module-script1.sh
                    \-another-module-script2.py

and the feature needs to be enabled by setting the following flag in the nextflow.config file. 

```
nextflow.enable.moduleBinaries = true
```

Note: it requires the use of a local or shared file system. The support for cloud native storage will be provided soon.

**Comment by replikation** [comment]
this is great thank you
